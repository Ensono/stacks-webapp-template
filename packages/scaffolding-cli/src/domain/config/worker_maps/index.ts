import * as shared from './shared'
import * as ssr from './ssr_aks_tfs'
import * as gkeSsr from './ssr_gke_tfs'
import * as gkeSsrJenkins from './ssr_gke_jenkins'
import * as netcore from './netcore_aks_tfs'
import * as javaSpring from './java_spring_aks_tfs'
import * as csr from './csr_aks_tfs'
import * as netcoreSelenium from './netcore_selenium_tfs'
import * as infraAks from './infra_aks_tfs'
import * as infraGke from './infra_gke_tfs'
import * as infraGkeJenkins from './infra_gke_jenkins'
import * as jsTestcafe from './js_testcafe_tfs'

export {
    ssr,
    gkeSsrJenkins,
    gkeSsr,
    netcore,
    javaSpring,
    csr,
    shared,
    netcoreSelenium,
    infraAks,
    infraGke,
    infraGkeJenkins,
    jsTestcafe
}
