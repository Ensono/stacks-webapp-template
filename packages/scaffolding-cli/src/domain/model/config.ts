export interface FolderMap {
    src: string
    dest: string
}

export interface SingleConfigKey {
    gitRepo: string
    gitRef: string
    folderMap: Array<FolderMap>
    localPath: string
    searchValue?: string
}

export interface Static {
    [key: string]: SingleConfigKey
}
