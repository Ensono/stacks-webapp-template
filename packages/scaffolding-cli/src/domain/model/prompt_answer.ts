export interface BusinessSection {
    company: string
    project: string
    domain: string
    component?: string
}

export const PlatformTypeEnum = {
    AKS: 'aks',
    GKE: 'gke',
    EKS: 'eks',
    ANY: 'any'
} as const;

export const DeploymentTypeEnum = {
    TFS: 'azdevops',
    JENKINS: 'jenkins'
} as const;


export interface CloudSection {
    region: string
    resourceGroup?: string
}

export interface SourceControlSection {
    repoType?: string
    repoName: string
    repoUrl?: string
}

export interface NetworkingSection {
    baseDomain: string
}


export interface TerraformSection {
    backendStorage: string
    backendStorageRg?: string
    backendStorageContainer?: string
}
export interface BaseAnswer {
    businessCompany: string
    projectName: string
    projectType: string
    businessDomain: string
    deployment: typeof DeploymentTypeEnum[keyof typeof DeploymentTypeEnum]
    platform?: typeof PlatformTypeEnum[keyof typeof PlatformTypeEnum]
}

/**
 * extend these with more optional strings
 * these should all be strings and optional
 */
export interface PromptAnswer extends BaseAnswer {
    enableAdvanced?: boolean
    businessComponent?: string
    cloudRegion?: string
    cloudResourceGroup?: string
    sourceControlRepoUrl?: string
    sourceControlRepoName?: string
    sourceControlRepoType?: string
    terraformBackendStorage?: string
    terraformBackendStorageRg?: string
    terraformBackendStorageContainer?: string
    networkingBaseDomain?: string
}

export interface CliAnswerModel extends BaseAnswer {
    enableAdvanced: boolean
    business: BusinessSection
    cloud: CloudSection
    terraform: TerraformSection
    sourceControl: SourceControlSection
    networking: NetworkingSection
    // add more here if needed
}
