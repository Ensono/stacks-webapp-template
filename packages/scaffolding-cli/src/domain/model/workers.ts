export interface CliError {
    message: string
    stack: string
}

export interface BaseResponse { 
    ok: boolean,
    message: string,
    code?: number | string
    error?: CliError
}

export interface TempCopy extends BaseResponse{
    tempPath: string
    finalPath: string
}

export interface ConfigResponse extends BaseResponse{
    configPath: string
}

export type CliResponse = BaseResponse
