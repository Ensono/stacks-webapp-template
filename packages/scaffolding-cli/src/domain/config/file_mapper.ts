// File Mapper
import { map, forEach, keys, forOwn, get, join } from 'lodash'
import { resolve } from 'path'
import { CliOptions } from '../model/cli_response';
import { CliAnswerModel } from '../model/prompt_answer';
import { isObject, isString } from 'util';

export enum BaseFlowType {
    BUILD = "build",
    SOURCE = "src",
    DEPLOY = "deploy",
    DOCS = "docs"   
}

export interface FileMapper {
    name: string,
    type: BaseFlowType
}

export interface ReplaceValMap {
    from: string | RegExp
    to: string
}

export interface BuildReplaceInputValues {
    [key: string]: string
}
export interface BuildReplaceInput {
    files: Array<string>
    values: BuildReplaceInputValues
    // [{files: ["README.md"], values: {find: "PROJECT_NAME", replace: instructions.project_name }}]
}

export interface Replacetruct {
    replaceFiles: Array<string>
    replaceVals: ReplaceValMap
    ignoreFiles?: Array<string>
    countMatches?: boolean
}

export function buildReplaceFoldersAndVals(base_path: string, replaceInput: Array<BuildReplaceInput>, ignoreFiles?: Array<string>, countMatches?: boolean): Array<Replacetruct> {
    let replaceOutput: Array<Replacetruct> = new Array<Replacetruct>()
    
    forEach(replaceInput, (input: BuildReplaceInput) => {
        forEach(keys(input.values), val => {
            // Replacetruct
            replaceOutput = [...replaceOutput, <Replacetruct>{
                replaceFiles: map(input.files, (v) => {return resolve(base_path, v)}),
                replaceVals: <ReplaceValMap>{
                    from: new RegExp(val, 'gi'),
                    to: input.values[val]
                }
            }]
        });
    })
    return replaceOutput
}

export function replaceGeneratedConfig(base_file: string, replaceInput: CliAnswerModel, ignoreFiles?: Array<string>, countMatches?: boolean): Array<Replacetruct> {
    let replaceOutput: Array<Replacetruct> = new Array<Replacetruct>()
    let objectKeys: Array<string> = Object.keys(replaceInput)
    let cheatConfig: any = replaceInput as any
    // forEach(keys(replaceInput), (input: typeof keys CliAnswerModel) => {
    function closureStruct (replaceOutput: Array<Replacetruct>, replaceVal: string, propertyPath: string): Replacetruct {
        return <Replacetruct>{
            replaceFiles: [base_file], // map(input.files, (v) => {return resolve(base_path, v)}),
            replaceVals: <ReplaceValMap>{
                from: new RegExp(`replace_${replaceVal}`, 'gi'),
                to: get(cheatConfig, propertyPath)
            }
        }
    } 
    
    objectKeys.forEach((input: any) => {
        if (isObject(cheatConfig[input])) {
            Object.keys(cheatConfig[input]).forEach((deepKey: any) => {
                replaceOutput = [...replaceOutput, closureStruct(replaceOutput, `${input}_${deepKey}`, `${input}.${deepKey}`)]
            })
        } else {
            replaceOutput = [...replaceOutput, closureStruct(replaceOutput, input, input)]
        }
    })
    return replaceOutput
}
