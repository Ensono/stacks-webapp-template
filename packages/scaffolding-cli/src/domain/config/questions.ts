import { PromptQuestion } from "../model/prompt_question"
import { PromptAnswer, CliAnswerModel, ProjectTypeEnum } from "../model/prompt_answer"

export function computedSelection(
    cliOrConfigAnswer: PromptAnswer,
): CliAnswerModel {
    return {
        projectName: cliOrConfigAnswer.projectName,
        projectType: cliOrConfigAnswer.projectType,
        deployment: cliOrConfigAnswer.deployment,
        platform: cliOrConfigAnswer.platform || "any",
        enableAdvanced: cliOrConfigAnswer.enableAdvanced || false,
        business: {
            company: cliOrConfigAnswer.businessCompany,
            project: cliOrConfigAnswer.projectName,
            domain: cliOrConfigAnswer.businessDomain,
            component:
                cliOrConfigAnswer.businessComponent || "COMPONENT_REPLACE_ME",
        },
        cloud: {
            region: cliOrConfigAnswer.cloudRegion || "REGION_REPLACE_ME",
        },
        terraform: {
            backendStorage:
                cliOrConfigAnswer.terraformBackendStorage ||
                "BACKEND_STORAGE_REPLACE_ME",
            backendStorageRg:
                cliOrConfigAnswer.terraformBackendStorageRg ||
                "BACKEND_STORAGE_RG_REPLACE_ME",
            backendStorageContainer: 
                cliOrConfigAnswer.terraformBackendStorageContainer ||
                "BACKEND_STORAGE_CONTAINER_REPLACE_ME",
        },
        sourceControl: {
            repoName:
                cliOrConfigAnswer.sourceControlRepoName ||
                "REPO_NAME_REPLACE_ME",
        },
        networking: {
            baseDomain:
                cliOrConfigAnswer.networkingBaseDomain ||
                "BASE_DOMAIN_NAME_REPLACE_ME",
        },
        javaspring: {
            namespace: cliOrConfigAnswer.javaTldNamespace ||
            "com"
        }

    } as CliAnswerModel
}

export const sharedInitialQs = (defaultProjectName: string): Array<PromptQuestion> => {
    return [{
        type: "text",
        name: "businessCompany",
        message: "Please provide the company name",
        initial: "amido",
    },
    {
        type: "text",
        name: "projectName",
        message: "Please provide the project name",
        initial: defaultProjectName,
    }]
}

export const sharedPostQs = (): Array<PromptQuestion> => {
    return [
        {
            type: "text",
            name: "businessDomain",
            message: "Please provide scope (domain)",
            description: "Used for scope with framework naming conventions.",
            initial: "menu-api",
        },
        {
            type: "select",
            name: "deployment",
            message: "Select Pipeline Tool",
            choices: [
                {
                    title: "AzureDevOps",
                    description: "Azure Devops/VSTS/TFS",
                    value: "azdevops",
                },
                {
                    title: "Jenkins",
                    description: "Jenkins CI/CD",
                    value: "jenkins",
                },
            ],
            initial: 0,
        }
    ]
}

/**
 * Additional questions should be placed here as we are extending the program
 */
export function cliQuestions(
    defaultProjectName: string,
): Array<PromptQuestion> {

    // let questions: Array<PromptQuestion> = Array<PromptQuestion
    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
    const custom: PromptQuestion = <PromptQuestion>{
        type: "select",
        name: "projectType",
        message: "Select Project type",
        choices: [
            {
                title: "React app with server side rendering",
                description:
                    "React, SSR, node, next.js, express, typescript",
                value: "ssr",
            },
            {
                title: "React app with client side rendering",
                description: "React, CSR, typescript",
                value: "csr",
            },
            {
                title: "API with .NET",
                description: "api, netcore, server, restfull",
                value: "netcore",
            },
            {
                title: "API with Java",
                description: "api, java, springboot",
                value: "javaspring",
            },
            {
                title: "Cloud platform shared services",
                description:
                    "terraform, azure, gcp, gke, aks, azure devops (tfs), jenkins",
                value: "infra",
            },
        ],
        initial: 0,
    }

    const questions: Array<PromptQuestion> = Array<PromptQuestion>().concat(sharedInitialQs(defaultProjectName), [custom], sharedPostQs(), platformQuestions())
    return questions
}

export function cliTestQuestions(
    defaultProjectName: string,
): Array<PromptQuestion> {
    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
    const custom: PromptQuestion = <PromptQuestion>{
        type: "select",
        name: "projectType",
        message: "Select test framework",
        choices: [
            {
                title: "Selenium framework with .NET",
                description:
                    "automation, bddfy, xunit, webdriver, netcore, e2e, standalone",
                value: "testnetcoreselenium",
            },
            {
                title: "TestCafe framework with Typescript",
                description:
                    "in-browser, automation, node.js, javascript, bdd, cross browser",
                value: "testjstestcafe",
            }
        ],
        initial: 0,
    }
    const questions: Array<PromptQuestion> = Array<PromptQuestion>().concat(sharedInitialQs(defaultProjectName), [custom], sharedPostQs())
    return questions
}

export function advancedQuestions(): Array<PromptQuestion> {
    return [
        {
            type: "text",
            name: "sourceControlRepoName",
            message: "Please provide version control repository name",
            initial: "stacks-repo",
        },
        {
            type: "text",
            name: "cloudRegion",
            message: "Please provide platform service region",
            initial: "uksouth",
        },
        {
            type: "select",
            name: "terraformBackendStorage",
            message: "Select Terraform remote state storage service?",
            choices: [
                {
                    title: "Azure",
                    description: "Azure Blob",
                    value: "azureBlob",
                },
                {
                    title: "AWS S3",
                    description: "S3 bucket",
                    value: "awsS3",
                },
                {
                    title: "GCS",
                    description: "GCP Storage bucket",
                    value: "gcpGcs",
                },
            ],
            initial: 0,
        },
        {
            type: "text",
            name: "networkingBaseDomain",
            message: "Provide host name (DNS domain)",
            initial: "app.replaceme.com",
        },
    ]
}

export function platformQuestions(): Array<PromptQuestion> {
    return [
        {
            type: "select",
            name: "platform",
            message: "Select Target Cloud Platform",
            choices: [
                {
                    title: "AKS",
                    description: "Azure Kubernetes",
                    value: "aks",
                },
                {
                    title: "GKE",
                    description: "Google Kubernetes Engine",
                    value: "gke",
                },
            ],
            initial: 0,
        },
        {
            type: "confirm",
            name: "enableAdvanced",
            message: "Continue to additional project configuration?",
            initial: true,
        },
    ]
}

export const language: {[key in ProjectTypeEnum]?: Function} = {
    [ProjectTypeEnum.JAVASPRING]: javaQuestions
}

export const platform: {[key: string]: Function} = {
}

export function javaQuestions(): Array<PromptQuestion> {
    return [
        {
            type: "text",
            name: "javaTldNamespace",
            message: "Package prefix - will be prepended to companyName.projectName previously supplied",
            initial: "org",
        }
    ]
}
