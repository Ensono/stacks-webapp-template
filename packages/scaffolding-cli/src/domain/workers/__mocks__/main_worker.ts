import { PromptAnswer } from '../../model/prompt_answer'
import { CliResponse, CliError } from '../../model/workers'

export class MainWorker {
    /**
     * 
     * @param instructions 
     */
    async ssr_aks_tfs(instructions: PromptAnswer): Promise<any> {
        let mockRes: CliResponse = {
            ok: true,
            message: "" 
        } as CliResponse
        try {
            return await jest.fn(() => mockRes)
        } catch (ex) {
            return ex;
        }
    }
}

export default {
    MainWorker
}
