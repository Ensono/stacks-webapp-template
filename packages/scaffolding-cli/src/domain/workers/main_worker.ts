import { CliAnswerModel } from '../model/prompt_answer'
import { CliResponse, CliError, TempCopy } from '../model/workers'
import { Utils } from './utils'
import { Replacetruct, buildReplaceFoldersAndVals, BuildReplaceInput } from '../config/file_mapper'
import { ssr, netcore, java_spring, csr, shared } from '../config/worker_maps'
import conf from '../config/static.config.json'
import { Static } from '../model/config'

let staticConf: Static = conf as Static;

export class MainWorker {
    /**
     * performs and entire templated out solution for the SSR AKS TFS deployment
     * @param {CliAnswerModel} instructions 
     * @returns 
     */
    async ssr_aks_tfs(instructions: CliAnswerModel): Promise<CliResponse> {
        let selectedFlowResponse: CliResponse = <CliResponse>{}
        try {

            let sharedBuildInput: Array<BuildReplaceInput> = shared.in_files({
                project_name: instructions.project_name,
                business_obj: instructions.business,
                cloud_obj: instructions.cloud,
                terraform_obj: instructions.terraform,
                scm_obj: instructions.source_control
            })

            let buildInput: Array<BuildReplaceInput> = ssr.in_files({
                project_name: instructions.project_name,
                business_obj: instructions.business,
                cloud_obj: instructions.cloud,
                terraform_obj: instructions.terraform,
                scm_obj: instructions.source_control
            }).concat(sharedBuildInput)

            let new_directory: TempCopy = await Utils.prepBase(instructions.project_name)

            await Utils.constructOutput(staticConf.ssr.folder_map, new_directory.final_path, new_directory.temp_path)

            let val_maps: Array<Replacetruct> = buildReplaceFoldersAndVals(new_directory.final_path, buildInput);

            await Utils.valueReplace(val_maps)
            if (instructions.create_config) {
                await Utils.writeOutConfigFile(`${instructions.project_name}.bootstrap-config.json`, instructions)
            }
            selectedFlowResponse.code = 0
            selectedFlowResponse.ok = true
            // Control the output message from each method
            selectedFlowResponse.message = shared.final_response_message(instructions.project_name, ssr.response_message(instructions.project_name), instructions.create_config)
            return selectedFlowResponse
        } catch (ex) {
            const cliErr = ex as CliError
            return <CliResponse>{
                ok: false,
                code: ex.code || -1,
                message: ex.message,
                error: cliErr
            };
        }
    }
    async netcore_aks_tfs(instructions: CliAnswerModel): Promise<CliResponse> {
        let selectedFlowResponse: CliResponse = <CliResponse>{}
        try {

            let sharedBuildInput: Array<BuildReplaceInput> = shared.in_files({
                project_name: instructions.project_name,
                business_obj: instructions.business,
                cloud_obj: instructions.cloud,
                terraform_obj: instructions.terraform,
                scm_obj: instructions.source_control
            })

            let buildInput: Array<BuildReplaceInput> = netcore.in_files({
                project_name: instructions.project_name,
                business_obj: instructions.business,
                cloud_obj: instructions.cloud
            }).concat(sharedBuildInput)

            let new_directory: TempCopy = await Utils.prepBase(instructions.project_name)
            // git clone node_repo custom app src
            // src_path_in_tmp should be statically defined in each method
            await Utils.doGitClone(staticConf.netcore.git_repo, new_directory.temp_path, staticConf.netcore.local_path, staticConf.netcore.git_ref)

            await Utils.constructOutput(staticConf.netcore.folder_map, new_directory.final_path, new_directory.temp_path)

            let val_maps: Array<Replacetruct> = buildReplaceFoldersAndVals(new_directory.final_path, buildInput)

            await Utils.valueReplace(val_maps)
            await Utils.fileNameReplace([`${new_directory.final_path}/src`, `${new_directory.final_path}/test`], instructions)
            if (instructions.create_config) {
                await Utils.writeOutConfigFile(`${instructions.project_name}.bootstrap-config.json`, instructions)
            }
            selectedFlowResponse.code = 0
            selectedFlowResponse.ok = true
            // Control the output message from each method
            selectedFlowResponse.message = shared.final_response_message(instructions.project_name, netcore.response_message(instructions.project_name), instructions.create_config)

            return selectedFlowResponse
        } catch (ex) {
            const cliErr = ex as CliError
            return <CliResponse>{
                ok: false,
                code: ex.code || -1,
                message: ex.message,
                error: cliErr
            };
        }
    }
    async java_spring_aks_tfs(instructions: CliAnswerModel): Promise<CliResponse> {
        let selectedFlowResponse: CliResponse = <CliResponse>{}
        try {

            let buildInput: Array<BuildReplaceInput> = java_spring.in_files(instructions.project_name, instructions.business, instructions.cloud)

            let new_directory: TempCopy = await Utils.prepBase(instructions.project_name)
            // git clone node_repo custom app src
            // src_path_in_tmp should be statically defined in each method
            await Utils.doGitClone(staticConf.java_spring.git_repo, new_directory.temp_path, staticConf.java_spring.local_path, staticConf.java_spring.git_ref)

            await Utils.constructOutput(staticConf.java_spring.folder_map, new_directory.final_path, new_directory.temp_path)

            let val_maps: Array<Replacetruct> = buildReplaceFoldersAndVals(new_directory.final_path, buildInput)

            await Utils.valueReplace(val_maps)

            if (instructions.create_config) {
                await Utils.writeOutConfigFile(`${instructions.project_name}.bootstrap-config.json`, instructions)
            }
            selectedFlowResponse.code = 0
            selectedFlowResponse.ok = true
            // Control the output message from each method
            selectedFlowResponse.message = shared.final_response_message(instructions.project_name, java_spring.response_message(instructions.project_name), instructions.create_config)

            return selectedFlowResponse
        } catch (ex) {
            const cliErr = ex as CliError
            return <CliResponse>{
                ok: false,
                code: ex.code || -1,
                message: ex.message,
                error: cliErr
            };
        }
    }
    async csr_aks_tfs(instructions: CliAnswerModel): Promise<CliResponse> {
        let selectedFlowResponse: CliResponse = <CliResponse>{}
        try {

            let buildInput: Array<BuildReplaceInput> = csr.in_files(instructions.project_name, instructions.business, instructions.cloud)

            let new_directory: TempCopy = await Utils.prepBase(instructions.project_name)
            // git clone node_repo custom app src
            // src_path_in_tmp should be statically defined in each method
            await Utils.doGitClone(staticConf.csr.git_repo, new_directory.temp_path, staticConf.csr.local_path, staticConf.csr.git_ref)

            await Utils.constructOutput(staticConf.csr.folder_map, new_directory.final_path, new_directory.temp_path)

            let val_maps: Array<Replacetruct> = buildReplaceFoldersAndVals(new_directory.final_path, buildInput)

            await Utils.valueReplace(val_maps)

            if (instructions.create_config) {
                await Utils.writeOutConfigFile(`${instructions.project_name}.bootstrap-config.json`, instructions)
            }

            selectedFlowResponse.code = 0
            selectedFlowResponse.ok = true
            // Control the output message from each method
            selectedFlowResponse.message = shared.final_response_message(instructions.project_name, csr.response_message(instructions.project_name), instructions.create_config)
            return selectedFlowResponse
        } catch (ex) {
            const cliErr = ex as CliError
            return <CliResponse>{
                ok: false,
                code: ex.code || -1,
                message: ex.message,
                error: cliErr
            };
        }
    }
}

export default {
    MainWorker
}
