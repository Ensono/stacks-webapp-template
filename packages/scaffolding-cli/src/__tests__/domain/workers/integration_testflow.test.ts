/* eslint-disable no-restricted-syntax */
import {promises, remove, ensureDir, emptyDir} from "fs-extra"
import path from "path"
import {CliAnswerModel} from "../../../domain/model/prompt_answer"
import {CliResponse} from "../../../domain/model/workers"
import {MainWorker} from "../../../domain/workers/main_worker"
import {jsTestcafe, netcoreSelenium} from "../../../domain/config/worker_maps"

const mockAnswer = {
    projectName: "testProjectName",
    projectType: "testProjectType",
    platform: "aks",
    deployment: "azdevops",
    business: {
        company: "testComp",
        domain: "testDomain",
        project: "testProject",
    },
} as CliAnswerModel

const mainWorker = new MainWorker()

jest.setTimeout(30000)

// Todo: extra this out into test util
// Credit: https://gist.github.com/lovasoa/8691344#gistcomment-3299089
async function* walk(dir: string): AsyncGenerator<string> {
    for await (const d of await promises.opendir(dir)) {
        const entry = path.join(dir, d.name)
        if (d.isDirectory()) yield* walk(entry)
        else if (d.isFile()) yield entry.replace(process.cwd(), "")
    }
}

describe("mainWorker class", () => {
    const currentDir = process.cwd()
    let tempDir: string

    beforeEach(async () => {
        tempDir = path.join(__dirname, mockAnswer.projectName)
        try {
            await ensureDir(tempDir)
            await emptyDir(tempDir)
            process.chdir(tempDir)
        } catch (err) {
            console.log(`chdir: ${err}`)
        }
    })

    afterEach(async () => {
        process.chdir(currentDir)
        await remove(tempDir)
    })

    it("netcoreSeleniumTfs with correct files", async () => {
        const results: string[] = []

        const flowRan: CliResponse = await mainWorker.netcoreSeleniumTfs(
            mockAnswer,
        )

        for await (const p of walk(tempDir)) results.push(p)

        expect(results.sort()).toMatchSnapshot()

        expect(flowRan.ok).toBe(true)
        expect(flowRan).toHaveProperty("message")

        expect(flowRan.message).toContain(
            netcoreSelenium.responseMessage(mockAnswer.projectName),
        )
    })

    it("jsTestcafe with correct files", async () => {
        const results: string[] = []

        const flowRan: CliResponse = await mainWorker.jsTestcafeTfs(mockAnswer)

        for await (const p of walk(tempDir)) results.push(p)

        expect(results.sort()).toMatchSnapshot()

        expect(flowRan.ok).toBe(true)
        expect(flowRan).toHaveProperty("message")

        expect(flowRan.message).toContain(
            jsTestcafe.responseMessage(mockAnswer.projectName),
        )
    })
})
