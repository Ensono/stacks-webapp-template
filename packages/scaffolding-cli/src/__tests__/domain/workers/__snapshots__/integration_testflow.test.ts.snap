// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mainWorker class jsTestcafe with correct files 1`] = `
Array [
  "/testProjectName.bootstrap-config.json",
  "/testProjectName/.eslintrc",
  "/testProjectName/.testcaferc.json",
  "/testProjectName/README.md",
  "/testProjectName/api/menu.ts",
  "/testProjectName/build/azDevops/azure/azure-pipeline-post-deploy-testcafe.yml",
  "/testProjectName/environment-variables.ts",
  "/testProjectName/fixtures/get-menu.test.cf.ts",
  "/testProjectName/fixtures/yumido.test.axe.cf.ts",
  "/testProjectName/package-lock.json",
  "/testProjectName/package.json",
  "/testProjectName/tsconfig.testcafe.json",
  "/testProjectName/utils/utils.cf.ts",
]
`;

exports[`mainWorker class netcoreSeleniumTfs with correct files 1`] = `
Array [
  "/testProjectName.bootstrap-config.json",
  "/testProjectName/Api/Builders/Http/HttpRequestBuilder.cs",
  "/testProjectName/Api/Builders/Http/HttpRequestFactory.cs",
  "/testProjectName/Configuration/ConfigAccessor.cs",
  "/testProjectName/Configuration/ConfigModel.cs",
  "/testProjectName/README.md",
  "/testProjectName/Selenium/Components/CreateForm.cs",
  "/testProjectName/Selenium/Components/Header.cs",
  "/testProjectName/Selenium/Components/MenuList.cs",
  "/testProjectName/Selenium/Components/Notifier.cs",
  "/testProjectName/Selenium/Factory/WebDriverFactory.cs",
  "/testProjectName/Selenium/Pages/PageObject.cs",
  "/testProjectName/Selenium/SeleniumWrapper.cs",
  "/testProjectName/Testcomp.Testproject.E2E.Selenium.csproj",
  "/testProjectName/Testcomp.Testproject.E2E.Selenium.sln",
  "/testProjectName/Tests/Fixtures/BaseSetup.cs",
  "/testProjectName/Tests/Steps/CreateMenu.cs",
  "/testProjectName/Tests/Stories/GetMenuExampleTests.cs",
  "/testProjectName/appsettings.json",
]
`;
